# 4.8 Grouping data and aggregating variables

### This script contains the following points:

#### 1. Create crosstab for final from merged data
#### 2. Creating a loyalty flag for customers
#### 3. Creating a spending flag for customers
#### 4. Creating an order frequency flag for customers

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
# Path import

path = r'/Users/nekow/Documents/Instacart Basket Analysis'

In [3]:
# Importing pickle file

ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_analysis.pkl'))

In [4]:
# Grouping the products

ords_prods_merge.groupby('product_name')

## 1. Create crosstab for final from merged data

In [5]:
crosstab = pd.crosstab(ords_prods_merge['days_since_prior_order'], ords_prods_merge['order_number'], dropna = False)

In [6]:
crosstab.to_clipboard()

In [5]:
# Splitting the data into groups


ords_prods_merge.groupby('department_id').agg({'order_number':['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

### Step 3 of Exercise 4.8
The entire dataframe has an ordered numerical organization rather than that of the set with only part of the dataframe.

In [6]:
# Producing the mean, min, and max

ords_prods_merge.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
1                15.457838   1  99
2                17.277920   1  99
3                17.170395   1  99
4                17.811403   1  99
5                15.215751   1  99
6                16.439806   1  99
7                17.225802   1  99
8                15.340650   1  99
9                15.895474   1  99
10               20.197148   1  99
11               16.170638   1  99
12               15.887671   1  99
13               16.583536   1  99
14               16.773669   1  99
15               16.165037   1  99
16               17.665606   1  99
17               15.694469   1  99
18               19.310397   1  99
19               17.177343   1  99
20               16.473447   1  99
21               22.902379   1  99

In [7]:
ords_prods_merge.head()

Unnamed: 0  product_id                product_name  aisle_id  \
0           0           1  Chocolate Sandwich Cookies        61   
1           0           1  Chocolate Sandwich Cookies        61   
2           0           1  Chocolate Sandwich Cookies        61   
3           0           1  Chocolate Sandwich Cookies        61   
4           0           1  Chocolate Sandwich Cookies        61   

   department_id  prices  order_id  user_id eval_set  order_number  \
0             19     5.8   3139998      138    prior            28   
1             19     5.8   1977647      138    prior            30   
2             19     5.8    389851      709    prior             2   
3             19     5.8    652770      764    prior             1   
4             19     5.8   1813452      764    prior             3   

   orders_day_of_week  order_hour_of_day  days_since_prior_order  \
0                   6                 11                     3.0   
1                   6                 17                    20.0   
2                   0                 21                     6.0   
3                   3                 13                     NaN   
4                   4                 17                     9.0   

   add_to_cart_order  reordered _merge        price_range    price_range_loc  \
0                  5          0   both  Mid-range product  Mid-range product   
1                  1          1   both  Mid-range product  Mid-range product   
2                 20          0   both  Mid-range product  Mid-range product   
3                 10          0   both  Mid-range product  Mid-range product   
4                 11          1   both  Mid-range product  Mid-range product   

   Busiest_days busiest_period_of_day  
0  Average Days           Most orders  
1  Average Days        Average orders  
2  Busiest days        Average orders  
3  Slowest days           Most orders  
4  Slowest days        Average orders

In [8]:
# dropping unnecessary columns

ords_prods_merge = ords_prods_merge.drop(columns = ['price_range_loc'])

In [9]:
# Splitting the data into user_id columns

ords_prods_merge.groupby('user_id').agg({'order_number': ['max']})

order_number
                 max
user_id             
1                 10
2                 14
3                 12
4                  5
5                  4
...              ...
206205             3
206206            67
206207            16
206208            49
206209            13

[206209 rows x 1 columns]

### Step 4 of Exercise 4.8


In [10]:
# max order creation

ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

/var/folders/hs/djjm__mn4d33wxms4t4q0flh0000gp/T/ipykernel_5243/602137744.py:3: FutureWarning: The provided callable <function amax at 0x1131d0fe0> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)


In [11]:
# data check for 15

ords_prods_merge.head(15)

Unnamed: 0  product_id                product_name  aisle_id  \
0            0           1  Chocolate Sandwich Cookies        61   
1            0           1  Chocolate Sandwich Cookies        61   
2            0           1  Chocolate Sandwich Cookies        61   
3            0           1  Chocolate Sandwich Cookies        61   
4            0           1  Chocolate Sandwich Cookies        61   
5            0           1  Chocolate Sandwich Cookies        61   
6            0           1  Chocolate Sandwich Cookies        61   
7            0           1  Chocolate Sandwich Cookies        61   
8            0           1  Chocolate Sandwich Cookies        61   
9            0           1  Chocolate Sandwich Cookies        61   
10           0           1  Chocolate Sandwich Cookies        61   
11           0           1  Chocolate Sandwich Cookies        61   
12           0           1  Chocolate Sandwich Cookies        61   
13           0           1  Chocolate Sandwich Cookies        61   
14           0           1  Chocolate Sandwich Cookies        61   

    department_id  prices  order_id  user_id eval_set  order_number  \
0              19     5.8   3139998      138    prior            28   
1              19     5.8   1977647      138    prior            30   
2              19     5.8    389851      709    prior             2   
3              19     5.8    652770      764    prior             1   
4              19     5.8   1813452      764    prior             3   
5              19     5.8   1701441      777    prior            16   
6              19     5.8   1871483      825    prior             3   
7              19     5.8   1290456      910    prior            12   
8              19     5.8    369558     1052    prior            10   
9              19     5.8    589712     1052    prior            15   
10             19     5.8   1798115     1379    prior             1   
11             19     5.8    497193     1480    prior            36   
12             19     5.8   1727826     1480    prior            38   
13             19     5.8   2694698     1480    prior            39   
14             19     5.8   3171753     1494    prior             6   

    orders_day_of_week  order_hour_of_day  days_since_prior_order  \
0                    6                 11                     3.0   
1                    6                 17                    20.0   
2                    0                 21                     6.0   
3                    3                 13                     NaN   
4                    4                 17                     9.0   
5                    1                  7                    26.0   
6                    2                 14                    30.0   
7                    3                 10                    30.0   
8                    1                 20                    19.0   
9                    1                 12                    15.0   
10                   5                 11                     NaN   
11                   3                 13                     7.0   
12                   4                 13                     7.0   
13                   2                 12                    12.0   
14                   3                 10                     2.0   

    add_to_cart_order  reordered _merge        price_range  Busiest_days  \
0                   5          0   both  Mid-range product  Average Days   
1                   1          1   both  Mid-range product  Average Days   
2                  20          0   both  Mid-range product  Busiest days   
3                  10          0   both  Mid-range product  Slowest days   
4                  11          1   both  Mid-range product  Slowest days   
5                   7          0   both  Mid-range product  Busiest days   
6                   2          0   both  Mid-range product  Average Days   
7                   1          0   both  Mid-range product  Slowe

In [12]:
# removing limiters

pd.options.display.max_rows = None

In [13]:
# checking 100 columns

ords_prods_merge.head(100)

Unnamed: 0  product_id                product_name  aisle_id  \
0            0           1  Chocolate Sandwich Cookies        61   
1            0           1  Chocolate Sandwich Cookies        61   
2            0           1  Chocolate Sandwich Cookies        61   
3            0           1  Chocolate Sandwich Cookies        61   
4            0           1  Chocolate Sandwich Cookies        61   
5            0           1  Chocolate Sandwich Cookies        61   
6            0           1  Chocolate Sandwich Cookies        61   
7            0           1  Chocolate Sandwich Cookies        61   
8            0           1  Chocolate Sandwich Cookies        61   
9            0           1  Chocolate Sandwich Cookies        61   
10           0           1  Chocolate Sandwich Cookies        61   
11           0           1  Chocolate Sandwich Cookies        61   
12           0           1  Chocolate Sandwich Cookies        61   
13           0           1  Chocolate Sandwich Cookies        61   
14           0           1  Chocolate Sandwich Cookies        61   
15           0           1  Chocolate Sandwich Cookies        61   
16           0           1  Chocolate Sandwich Cookies        61   
17           0           1  Chocolate Sandwich Cookies        61   
18           0           1  Chocolate Sandwich Cookies        61   
19           0           1  Chocolate Sandwich Cookies        61   
20           0           1  Chocolate Sandwich Cookies        61   
21           0           1  Chocolate Sandwich Cookies        61   
22           0           1  Chocolate Sandwich Cookies        61   
23           0           1  Chocolate Sandwich Cookies        61   
24           0           1  Chocolate Sandwich Cookies        61   
25           0           1  Chocolate Sandwich Cookies        61   
26           0           1  Chocolate Sandwich Cookies        61   
27           0           1  Chocolate Sandwich Cookies        61   
28           0           1  Chocolate Sandwich Cookies        61   
29           0           1  Chocolate Sandwich Cookies        61   
30           0           1  Chocolate Sandwich Cookies        61   
31           0           1  Chocolate Sandwich Cookies        61   
32           0           1  Chocolate Sandwich Cookies        61   
33           0           1  Chocolate Sandwich Cookies        61   
34           0           1  Chocolate Sandwich Cookies        61   
35           0           1  Chocolate Sandwich Cookies        61   
36           0           1  Chocolate Sandwich Cookies        61   
37           0           1  Chocolate Sandwich Cookies        61   
38           0           1  Chocolate Sandwich Cookies        61   
39           0           1  Chocolate Sandwich Cookies        61   
40           0           1  Chocolate Sandwich Cookies        61   
41           0           1  Chocolate Sandwich Cookies        61   
42           0           1  Chocolate Sandwich Cookies        61   
43           0           1  Chocolate Sandwich Cookies        61   
44           0           1  Chocolate Sandwich Cookies        61   
45           0           1  Chocolate Sandwich Cookies        61   
46           0           1  Chocolate Sandwich Cookies        61   
47           0           1  Chocolate Sandwich Cookies        61   
48           0           1  Chocolate Sandwich Cookies        61   
49           0           1  Chocolate Sandwich Cookies        61   
50           0           1  Chocolate Sandwich Cookies        61   
51           0           1  Chocolate Sandwich Cookies        61   
52           0           1  Chocolate Sandwich Cookies        61   
53           0           1  Chocolate Sandwich Cookies        61   
54           0           1  Chocolate Sandwich Cookies        61   
55           0           1  Chocolate Sandwich Cookies        61   
56           0           1  Chocolate Sandwich Cookies        61   
57           0           1  Chocolate Sandwich Cookies      

In [14]:
# Dropping the unnamed columns

ords_prods_merge = ords_prods_merge.drop(columns = ['Unnamed: 0'])

## 2. Creating a loyalty flag for customers

In [15]:
# max order for loyal customers

ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

/var/folders/hs/djjm__mn4d33wxms4t4q0flh0000gp/T/ipykernel_5243/430384142.py:3: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value 'Loyal customer' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.
  ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'


In [16]:
# max order for regular customers

ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [17]:
# max order for new customers

ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [18]:
# value count for loyalty flag creation

ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [19]:
# Confirmation of loyalty flag creation

ords_prods_merge[['user_id', 'loyalty_flag', 'order_number']].head(15)

user_id      loyalty_flag  order_number
0       138  Regular customer            28
1       138  Regular customer            30
2       709      New customer             2
3       764      New customer             1
4       764      New customer             3
5       777  Regular customer            16
6       825      New customer             3
7       910  Regular customer            12
8      1052  Regular customer            10
9      1052  Regular customer            15
10     1379      New customer             1
11     1480    Loyal customer            36
12     1480    Loyal customer            38
13     1480    Loyal customer            39
14     1494  Regular customer             6

In [20]:
# Loyalty flag for customer spending habits basic statistics

ords_prods_merge.groupby('loyalty_flag').agg({'prices' : ['mean','min','max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

### Step 5 of Exercise 4.8
The mean suggests that New Customers spend more on average than any other customers, and Loyal Customers spend an average lower than even Regular Customers. All customers spend a lowest amount of 1 and a maximum amount of 99,999.

## 3. Creating a spending flag for customers

### Step 6 of Exercise 4.8

In [21]:
# column reation for the mean of prices

ords_prods_merge['mean_prices'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

/var/folders/hs/djjm__mn4d33wxms4t4q0flh0000gp/T/ipykernel_5243/1060439922.py:3: FutureWarning: The provided callable <function mean at 0x1131d1d00> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  ords_prods_merge['mean_prices'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)


In [22]:
# Data check

ords_prods_merge.head(15)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   

    order_id  user_id eval_set  order_number  orders_day_of_week  ...  \
0    3139998      138    prior            28                   6  ...   
1    1977647      138    prior            30                   6  ...   
2     389851      709    prior             2                   0  ...   
3     652770      764    prior             1                   3  ...   
4    1813452      764    prior             3                   4  ...   
5    1701441      777    prior            16                   1  ...   
6    1871483      825    prior             3                   2  ...   
7    1290456      910    prior            12                   3  ...   
8     369558     1052    prior            10                   1  ...   
9     589712     1052    prior            15                   1  ...   
10   1798115     1379    prior             1                   5  ...   
11    497193     1480    prior            36                   3  ...   
12   1727826     1480    prior            38                   4  ...   
13   2694698     1480    prior            39                   2  ...   
14   3171753     1494    prior             6                   3  ...   

    days_since_prior_order  add_to_cart_order  reordered  _merge  \
0                      3.0                  5          0    both   
1                     20.0                  1          1    both   
2                      6.0                 20          0    both   
3                      NaN                 10          0    both   
4                      9.0                 11          1    both   
5                     26.0                  7          0    both   
6                     30.0                  2          0    both   
7                     30.0                  1          0    both   
8                     19.0                  1          0    both   
9                     15.0                  2          1    both   
10                     NaN                  1          0    both   
11                     7.0                  1          0    both   
12                     7.0                  3          1    both   
13                    12.0                  4          1    both   
14                     2.0                  8          0    both   

          price_range  Busiest_days busiest_period_of_day max_order  \
0   Mid-range product  Average Days           Most orders        32   
1   Mid-range product  Average Days        Average orders        32   
2   Mid-range product  Busiest days        Average orders         5   
3   Mid-range product  Slowest days           Most orders         3   
4   Mid-range product  Slowest days        Average orders         3   
5   Mid-range product  Busiest days        Average orde

In [23]:
# Low spender indicator

ords_prods_merge.loc[ords_prods_merge['mean_prices'] < 10, 'spending_flag'] = 'Low spender'

/var/folders/hs/djjm__mn4d33wxms4t4q0flh0000gp/T/ipykernel_5243/2461071344.py:3: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value 'Low spender' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.
  ords_prods_merge.loc[ords_prods_merge['mean_prices'] < 10, 'spending_flag'] = 'Low spender'


In [24]:
# High spender indicator

ords_prods_merge.loc[ords_prods_merge['mean_prices'] >= 10, 'spending_flag'] = 'High spender'

In [25]:
# flag consistenct check

ords_prods_merge[['user_id', 'spending_flag', 'product_id']].head(15)

user_id spending_flag  product_id
0       138   Low spender           1
1       138   Low spender           1
2       709   Low spender           1
3       764   Low spender           1
4       764   Low spender           1
5       777   Low spender           1
6       825   Low spender           1
7       910   Low spender           1
8      1052   Low spender           1
9      1052   Low spender           1
10     1379   Low spender           1
11     1480   Low spender           1
12     1480   Low spender           1
13     1480   Low spender           1
14     1494   Low spender           1

In [26]:
# Value count for spending flag

ords_prods_merge['spending_flag'].value_counts(dropna = False)

spending_flag
Low spender     31770614
High spender      634245
Name: count, dtype: int64

## 4. Creating an order frequency flag for customers

### Step 7 of Exercise 4.8

In [27]:
# median order frequency creation

ords_prods_merge['median_order_frequency'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

/var/folders/hs/djjm__mn4d33wxms4t4q0flh0000gp/T/ipykernel_5243/2457283246.py:3: FutureWarning: The provided callable <function median at 0x1133200e0> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  ords_prods_merge['median_order_frequency'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)


In [28]:
# data check

ords_prods_merge.tail(15)

product_id            product_name  aisle_id  department_id  prices  \
32404844       49688  Fresh Foaming Cleanser        73             11    13.5   
32404845       49688  Fresh Foaming Cleanser        73             11    13.5   
32404846       49688  Fresh Foaming Cleanser        73             11    13.5   
32404847       49688  Fresh Foaming Cleanser        73             11    13.5   
32404848       49688  Fresh Foaming Cleanser        73             11    13.5   
32404849       49688  Fresh Foaming Cleanser        73             11    13.5   
32404850       49688  Fresh Foaming Cleanser        73             11    13.5   
32404851       49688  Fresh Foaming Cleanser        73             11    13.5   
32404852       49688  Fresh Foaming Cleanser        73             11    13.5   
32404853       49688  Fresh Foaming Cleanser        73             11    13.5   
32404854       49688  Fresh Foaming Cleanser        73             11    13.5   
32404855       49688  Fresh Foaming Cleanser        73             11    13.5   
32404856       49688  Fresh Foaming Cleanser        73             11    13.5   
32404857       49688  Fresh Foaming Cleanser        73             11    13.5   
32404858       49688  Fresh Foaming Cleanser        73             11    13.5   

          order_id  user_id eval_set  order_number  orders_day_of_week  ...  \
32404844   1173882   178752    prior             7                   0  ...   
32404845   1525694   178752    prior            15                   5  ...   
32404846   2439069   182948    prior             4                   6  ...   
32404847   2392655   182948    prior             8                   2  ...   
32404848    848810   183548    prior            15                   2  ...   
32404849   2251059   184081    prior             1                   6  ...   
32404850    846716   187523    prior            14                   3  ...   
32404851    362546   188073    prior             5                   2  ...   
32404852   3166828   188073    prior             7                   1  ...   
32404853   3106363   197371    prior            10                   4  ...   
32404854   1788356   200215    prior             2                   0  ...   
32404855   3401313   200377    prior             1                   4  ...   
32404856    809510   200873    prior             5                   3  ...   
32404857   2359893   200873    prior             9                   3  ...   
32404858   2385091   205926    prior            11                   1  ...   

          reordered  _merge  price_range  Busiest_days busiest_period_of_day  \
32404844          0    both          NaN  Busiest days         Fewest orders   
32404845          1    both          NaN  Average Days           Most orders   
32404846          0    both          NaN  Average Days           Most orders   
32404847          1    both          NaN  Average Days           Most orders   
32404848          0    both          NaN  Average Days           Most orders   
32404849          0    both          NaN  Average Days           Most orders   
32404850          0    both          NaN  Slowest days           Most orders   
32404851          0    both          NaN  Average Days         Fewest orders   
32404852          1    both          NaN  Busiest days           Most orders   
32404853          0    both          NaN  Slowest days           Most orders   
32404854          0    both          NaN  Busiest days           Most orders   
32404855          0    both          NaN  Slowest days           Most orders   
32404856          0    both          NaN  Slowest days        Average orders   
32404857          1    both          NaN  Slowest days           Most orders   
32404858          0    both          NaN  Busiest days           Most orders   

         max_order      loyalty_flag mean_prices  spending_flag  \
32404844        16  Regular customer    8.643223    Low spender   
32404845        16  Regular custo

In [29]:
# median order for median frequency on non-frequent customers

ords_prods_merge.loc[ords_prods_merge['median_order_frequency'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'

/var/folders/hs/djjm__mn4d33wxms4t4q0flh0000gp/T/ipykernel_5243/1100560102.py:3: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value 'Non-frequent customer' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.
  ords_prods_merge.loc[ords_prods_merge['median_order_frequency'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'


In [30]:
# median order for median frequency on regular customers

ords_prods_merge.loc[(ords_prods_merge['median_order_frequency'] > 10) & (ords_prods_merge['median_order_frequency'] <= 20), 'order_frequency_flag'] = 'Regular customer'

In [31]:
# median order for median frequency on frequent customers

ords_prods_merge.loc[ords_prods_merge['median_order_frequency'] <= 10, 'order_frequency_flag'] = 'Frequent customer'

In [32]:
# Value count for order frequency flag

ords_prods_merge['order_frequency_flag'].value_counts(dropna = False)

order_frequency_flag
Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: count, dtype: int64

In [33]:
# finding the missing values for the new flag

ords_prods_merge[ords_prods_merge['order_frequency_flag'].isnull() == True]

product_id                                       product_name  \
6234909        10749                            Organic Red Bell Pepper   
12947653       21334                              Organic Peeled Garlic   
13839012       22198  4X Ultra Concentrated Natural Laundry Detergen...   
14758536       23695                           California Veggie Burger   
21673807       33401                               Goat Cheese Crumbles   

          aisle_id  department_id  prices  order_id  user_id eval_set  \
6234909         83              4     5.7    895835   159838    prior   
12947653       123              4    10.2    895835   159838    prior   
13839012        75             17     1.7    895835   159838    prior   
14758536        42              1     4.7    895835   159838    prior   
21673807        21             16    14.8    895835   159838    prior   

          order_number  orders_day_of_week  ...  _merge  price_range  \
6234909              1                   0  ...    both          NaN   
12947653             1                   0  ...    both          NaN   
13839012             1                   0  ...    both          NaN   
14758536             1                   0  ...    both          NaN   
21673807             1                   0  ...    both          NaN   

          Busiest_days  busiest_period_of_day max_order  loyalty_flag  \
6234909   Busiest days         Average orders         1  New customer   
12947653  Busiest days         Average orders         1  New customer   
13839012  Busiest days         Average orders         1  New customer   
14758536  Busiest days         Average orders         1  New customer   
21673807  Busiest days         Average orders         1  New customer   

         mean_prices spending_flag  median_order_frequency  \
6234909         7.42   Low spender                     NaN   
12947653        7.42   Low spender                     NaN   
13839012        7.42   Low spender                     NaN   
14758536        7.42   Low spender                     NaN   
21673807        7.42   Low spender                     NaN   

         order_frequency_flag  
6234909                   NaN  
12947653                  NaN  
13839012                  NaN  
14758536                  NaN  
21673807                  NaN  

[5 rows x 24 columns]

### There's a customer anomaly for 5 variables where they are a new customer with no prior purchases that doesn't fit the criteria for any frequency flags.

In [34]:
# data check

ords_prods_merge.tail(15)

product_id            product_name  aisle_id  department_id  prices  \
32404844       49688  Fresh Foaming Cleanser        73             11    13.5   
32404845       49688  Fresh Foaming Cleanser        73             11    13.5   
32404846       49688  Fresh Foaming Cleanser        73             11    13.5   
32404847       49688  Fresh Foaming Cleanser        73             11    13.5   
32404848       49688  Fresh Foaming Cleanser        73             11    13.5   
32404849       49688  Fresh Foaming Cleanser        73             11    13.5   
32404850       49688  Fresh Foaming Cleanser        73             11    13.5   
32404851       49688  Fresh Foaming Cleanser        73             11    13.5   
32404852       49688  Fresh Foaming Cleanser        73             11    13.5   
32404853       49688  Fresh Foaming Cleanser        73             11    13.5   
32404854       49688  Fresh Foaming Cleanser        73             11    13.5   
32404855       49688  Fresh Foaming Cleanser        73             11    13.5   
32404856       49688  Fresh Foaming Cleanser        73             11    13.5   
32404857       49688  Fresh Foaming Cleanser        73             11    13.5   
32404858       49688  Fresh Foaming Cleanser        73             11    13.5   

          order_id  user_id eval_set  order_number  orders_day_of_week  ...  \
32404844   1173882   178752    prior             7                   0  ...   
32404845   1525694   178752    prior            15                   5  ...   
32404846   2439069   182948    prior             4                   6  ...   
32404847   2392655   182948    prior             8                   2  ...   
32404848    848810   183548    prior            15                   2  ...   
32404849   2251059   184081    prior             1                   6  ...   
32404850    846716   187523    prior            14                   3  ...   
32404851    362546   188073    prior             5                   2  ...   
32404852   3166828   188073    prior             7                   1  ...   
32404853   3106363   197371    prior            10                   4  ...   
32404854   1788356   200215    prior             2                   0  ...   
32404855   3401313   200377    prior             1                   4  ...   
32404856    809510   200873    prior             5                   3  ...   
32404857   2359893   200873    prior             9                   3  ...   
32404858   2385091   205926    prior            11                   1  ...   

          _merge  price_range  Busiest_days  busiest_period_of_day max_order  \
32404844    both          NaN  Busiest days          Fewest orders        16   
32404845    both          NaN  Average Days            Most orders        16   
32404846    both          NaN  Average Days            Most orders         8   
32404847    both          NaN  Average Days            Most orders         8   
32404848    both          NaN  Average Days            Most orders        15   
32404849    both          NaN  Average Days            Most orders        11   
32404850    both          NaN  Slowest days            Most orders        17   
32404851    both          NaN  Average Days          Fewest orders        31   
32404852    both          NaN  Busiest days            Most orders        31   
32404853    both          NaN  Slowest days            Most orders        10   
32404854    both          NaN  Busiest days            Most orders         6   
32404855    both          NaN  Slowest days            Most orders         4   
32404856    both          NaN  Slowest days         Average orders        20   
32404857    both          NaN  Slowest days            Most orders        20   
32404858    both          NaN  Busiest days            Most orders        11   

              loyalty_flag mean_prices spending_flag  median_order_frequency  \
32404844  Regular customer    8.643223   Low spender                    21.0   
3240484

In [35]:
# flag consistenct check

ords_prods_merge[['user_id','order_frequency_flag','median_order_frequency']].head(15)

user_id order_frequency_flag  median_order_frequency
0       138    Frequent customer                     8.0
1       138    Frequent customer                     8.0
2       709    Frequent customer                     8.0
3       764    Frequent customer                     9.0
4       764    Frequent customer                     9.0
5       777     Regular customer                    11.0
6       825     Regular customer                    20.0
7       910    Frequent customer                     6.0
8      1052    Frequent customer                    10.0
9      1052    Frequent customer                    10.0
10     1379     Regular customer                    20.0
11     1480    Frequent customer                     7.0
12     1480    Frequent customer                     7.0
13     1480    Frequent customer                     7.0
14     1494    Frequent customer                     7.0

In [ ]:
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_flagged.pkl'))